File: Level3/정수 삼각형.cpp
#include <string>
#include <vector>

using namespace std;

//최댓값 반환
int MAX(int a, int b) {
    return a > b ? a : b;
}

vector<vector<int>> dp;
int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int size = triangle.size(); //길이
    dp.resize(size + 1, vector<int> (size + 1, 0)); //dp 초기화
    
    dp[1][1] = triangle[0][0];
    int max_value = dp[1][1];
    
    //보텀업방식
    for (int i = 2; i <= size; ++i) {
        //파스칼 규칙에 따라
        for (int j = 1; j<=i; ++j) {
            dp[i][j] = MAX(dp[i-1][j-1], dp[i-1][j]) + triangle[i - 1][j - 1];
            
            //최댓값 교체
            if (dp[i][j] > max_value)
                max_value = dp[i][j];
        }
    }
    
    answer = max_value; //최댓값
    return answer;
}

File: Level3/정수 삼각형.py
def solution(triangle):
    answer = 0

    #dp테이블 초기화
    dp = [[0] * len(triangle) for _ in range(len(triangle))]

    #첫번째 수 셋팅
    dp[0][0] = triangle[0][0]

    #보텀업
    # 거쳐간 숫자의 최댓값 구하기 
    for i in range(0, len(triangle) - 1): 
        for j in range(len(triangle[i])): 
            dp[i + 1][j] = max(dp[i + 1][j], dp[i][j] + triangle[i + 1][j]) 
            dp[i + 1][j + 1] = max(dp[i + 1][j + 1], dp[i][j] + triangle[i + 1][j + 1])

    #정수 삼각형 최하단줄에 큰값들이 있으므로
    answer = max(dp[-1])
    return answer